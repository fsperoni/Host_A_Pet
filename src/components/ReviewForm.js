import React, { useContext, useState } from 'react';
import UserContext from '../hooks/useUserContext';
import Alert from "./Alert";
import HostAPetApi from './Api';

const ReviewForm = ({ review, reviewee, setReview, setShowReviewForm }) => {
  const { currentUser } = useContext(UserContext);
  const [formData, setFormData] = useState({
    id: review.id || 0,
    reviewer: review.reviewer || currentUser.username,
    reviewee: review.reviewee || reviewee,
    rating: review.rating || "",
    comments: review.comments || ""
  });
  const [formErrors, setFormErrors] = useState([]);
  const [reviewSuccess, setReviewSuccess] = useState([]);

  /** Handle form submit */
  const handleSubmit = async (evt) => {
    evt.preventDefault();
    let data = { ...formData };
    try {
      if (data.id === 0) {
        delete data.id;
        const result = await HostAPetApi.addReview(currentUser.username, data);
        setReview([result])
      } else {
        const id = data.id;
        delete data.id;
        const result = await HostAPetApi.updateReview(currentUser.username, id, data);
        setReview([result])
      }
      setReviewSuccess(["Review successfully processed!"]);
      setFormErrors([]);
    } catch (err) {
      setFormErrors(err);
    }
  }

  /** Handle form field changes */
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData(data => ({ ...data, [name]: value }));
    setFormErrors([]);
  }

  /** Handle form cancel button */
  const handleCancel = (evt) => {
    evt.preventDefault();
    setShowReviewForm(false);
  }

  return (
    <div className="ReviewForm mt-3">
      <div className="container">
        <h2 className="mb-3">Your review</h2>
        <div className="card">
          <div className="card-body">
            <form onSubmit={handleSubmit}>
              <div className="form-group">
                <label>Rating:</label>
                <select name="rating" required className="form-select"
                  onChange={handleChange} value={formData.rating}>
                  <option hidden="" required value="">Select one</option>
                  <option value={5}>5</option>
                  <option value={4}>4</option>
                  <option value={3}>3</option>
                  <option value={2}>2</option>
                  <option value={1}>1</option>
                  <option value={0}>0</option>
                </select>
              </div>
              <div className="form-group">
                <label>Comments:</label>
                <textarea name="comments" className="form-control" onChange={handleChange}
                  value={formData.comments}
                />
              </div>
              {formErrors.length > 0 && <Alert type="danger" messages={formErrors} />}
              <div className="container mt-4">
                <button
                  className="btn btn-primary btn-block"
                  onClick={handleSubmit}
                >
                  Save
                </button>
                <button
                  className="btn btn-warning btn-block ms-4"
                  onClick={handleCancel}
                >
                  Cancel
                </button>
                <div className="alert-container">
                  {reviewSuccess.length > 0 && <Alert type="success" messages={reviewSuccess} />}
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default ReviewForm;