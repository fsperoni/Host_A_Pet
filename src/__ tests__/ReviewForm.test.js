import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import ReviewForm from "../components/ReviewForm";
import UserContext from "../hooks/useUserContext";

const {
  currentAdminUser,
  reviewee, review
} = require("./_testCommon");

// smoke test
it('renders without crashing', function () {
  render(
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <ReviewForm setReview={function () { }} review={review} 
          setShowReviewForm={function () {}} reviewee={reviewee()}/>
      </UserContext.Provider>
    </BrowserRouter>
  );
});

//snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <ReviewForm setReview={function () { }} review={review} 
          setShowReviewForm={function () {}} reviewee={reviewee()}/>
      </UserContext.Provider>
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should have 'Your review' in the component.", function () {
  const { getByText } = render(
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <ReviewForm setReview={function () { }} review={review} 
          setShowReviewForm={function () {}} reviewee={reviewee()}/>
      </UserContext.Provider>
    </BrowserRouter>
  );
  expect(getByText("Your review")).toBeInTheDocument();
});
