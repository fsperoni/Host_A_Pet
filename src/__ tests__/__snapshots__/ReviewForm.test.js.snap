// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ReviewForm mt-3"
  >
    <div
      class="container"
    >
      <h2
        class="mb-3"
      >
        Your review
      </h2>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label>
                Rating:
              </label>
              <select
                class="form-select"
                name="rating"
                required=""
              >
                <option
                  required=""
                  value=""
                >
                  Select one
                </option>
                <option
                  value="5"
                >
                  5
                </option>
                <option
                  value="4"
                >
                  4
                </option>
                <option
                  value="3"
                >
                  3
                </option>
                <option
                  value="2"
                >
                  2
                </option>
                <option
                  value="1"
                >
                  1
                </option>
                <option
                  value="0"
                >
                  0
                </option>
              </select>
            </div>
            <div
              class="form-group"
            >
              <label>
                Comments:
              </label>
              <textarea
                class="form-control"
                name="comments"
              />
            </div>
            <div
              class="container mt-4"
            >
              <button
                class="btn btn-primary btn-block"
              >
                Save
              </button>
              <button
                class="btn btn-warning btn-block ms-4"
              >
                Cancel
              </button>
              <div
                class="alert-container"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
